fabric-data-generation-api-v1/src/main/java/net/fabricmc/fabric/api/datagen/v1/provider/FabricAdvancementsProvider.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.datagen.v1.provider;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Set;
+import java.util.function.Consumer;
+
+import com.google.common.collect.Sets;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+import net.minecraft.advancement.Advancement;
+import net.minecraft.data.DataCache;
+import net.minecraft.data.DataProvider;
+import net.minecraft.util.Identifier;
+
+import net.fabricmc.fabric.api.datagen.v1.FabricDataGenerator;
+
+/**
+ * Extend this class and implement {@link FabricAdvancementsProvider#generateAdvancement}.
+ *
+ * <p>Register an instance of the class with {@link FabricDataGenerator#addProvider} in a {@link net.fabricmc.fabric.api.datagen.v1.DataGeneratorEntrypoint}
+ */
+public abstract class FabricAdvancementsProvider implements DataProvider {
+	private static final Gson GSON = (new GsonBuilder()).setPrettyPrinting().create();
+
+	protected final FabricDataGenerator dataGenerator;
+
+	protected FabricAdvancementsProvider(FabricDataGenerator dataGenerator) {
+		this.dataGenerator = dataGenerator;
+	}
+
+	/**
+	 * Use {@link Advancement.Task#build(Consumer, String)} to register an advancement to build.
+	 */
+	public abstract void generateAdvancement(Consumer<Advancement> consumer);
+
+	@Override
+	public void run(DataCache cache) throws IOException {
+		final Set<Identifier> identifiers = Sets.newHashSet();
+		final Set<Advancement> advancements = Sets.newHashSet();
+
+		generateAdvancement(advancements::add);
+
+		for (Advancement advancement : advancements) {
+			if (!identifiers.add(advancement.getId())) {
+				throw new IllegalStateException("Duplicate advancement " + advancement.getId());
+			}
+
+			DataProvider.writeToPath(GSON, cache, advancement.createTask().toJson(), getOutputPath(advancement));
+		}
+	}
+
+	private Path getOutputPath(Advancement advancement) {
+		return dataGenerator.getOutput().resolve("data/%s/advancements/%s.json".formatted(advancement.getId().getNamespace(), advancement.getId().getPath()));
```

Comment:
Should use the mod id and not the namespace of the advancement.

Reply:
Could also make this method (and similar ones if they exist) protected.

Reply:
The advancement should have been already created with the mod id?

Reply:
Yeah just checked, and that's correct, I should create it with the mod id.

Reply:
![image](https://user-images.githubusercontent.com/13494793/143782572-6ecb9207-daaf-4cb9-a040-0d60a12819e4.png)

I was using this method, but if the modid is not right, child advancements will use the `minecraft` namespace to the parent anyway. So no need to change the id indeed.

Reply:
Yeah, slightly annoying that it takes a string and not an ID, would be more effort than its worth to find a solution to that.